src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ApiConventionItem.cs

```diff
@@ -0,0 +1,199 @@
+            {
+                return false;
+            }
+
+            var methodParameters = methodInfo.GetParameters();
+            var conventionMethodParameters = conventionMethod.GetParameters();
+
+            for (var i = 0; i < conventionMethodParameters.Length; i++)
+            {
+                var conventionParameter = conventionMethodParameters[i];
+                if (conventionParameter.IsDefined(typeof(ParamArrayAttribute)))
+                {
+                    return true;
+                }
+
+                if (methodParameters.Length <= i)
+                {
+                    return false;
+                }
+
+                var nameMatchBehavior = GetNameMatchBehavior(conventionParameter);
+                var typeMatchBehavior = GetTypeMatchBehavior(conventionParameter);
+
+                if (!IsTypeMatch(methodParameters[i].ParameterType, conventionParameter.ParameterType, typeMatchBehavior) ||
+                    !IsNameMatch(methodParameters[i].Name, conventionParameter.Name, nameMatchBehavior))
+                {
+                    return false;
+                }
+            }
+
+            // Ensure convention has at least as many parameters as the method. params convention argument are handled
+            // inside the for loop.
+            return methodParameters.Length == conventionMethodParameters.Length;
+
+            ApiConventionNameMatchBehavior GetNameMatchBehavior(ICustomAttributeProvider attributeProvider)
+            {
+                var attributes = attributeProvider.GetCustomAttributes(inherit: false);
+                for (var i = 0; i < attributes.Length; i++)
+                {
+                    if (attributes[i] is ApiConventionNameMatchAttribute nameMatchAttribute)
+                    {
+                        return nameMatchAttribute.MatchBehavior;
+                    }
+                }
+
+                return ApiConventionNameMatchBehavior.Exact;
+            }
+
+            ApiConventionTypeMatchBehavior GetTypeMatchBehavior(ParameterInfo parameter)
+            {
+                var typeMatchAttribute = parameter.GetCustomAttribute<ApiConventionTypeMatchAttribute>(inherit: false);
+                return typeMatchAttribute?.MatchBehavior ?? ApiConventionTypeMatchBehavior.Exact;
+            }
+        }
+
+        internal static bool IsNameMatch(string name, string conventionName, ApiConventionNameMatchBehavior nameMatchBehavior)
+        {
+            switch (nameMatchBehavior)
+            {
+                case ApiConventionNameMatchBehavior.Any:
+                    return true;
+
+                case ApiConventionNameMatchBehavior.Exact:
+                    return string.Equals(name, conventionName, StringComparison.Ordinal);
+
+                case ApiConventionNameMatchBehavior.Prefix:
+                    return IsNameMatchPrefix();
+
+                case ApiConventionNameMatchBehavior.Suffix:
+                    return IsNameMatchSuffix();
```

Comment:
Why does `Exact` look at `name` and `conventionName` but the others don't?

Reply:
Oh, local functions. Derp. This is silly reason to use local functions.

Reply:
💯 points